Implement CodecFactoryOptions allowing clients to opt-in to Pretty encoders and Strict Decoders

stealthybox opened the pull request:
/kind api-change
@kubernetes/wg-component-standard 

**What this PR does / why we need it**:
This patch introduces a new CodecFactory constructor that's capable of building different collections of Serializers.
Note that this diverges from the current design where an initialized CodecFactory is a kitchen sink holding many forms of useful Serilaizers in a non-customizeable way.

This approach is a suggested solution from this thread:
https://groups.google.com/d/msg/kubernetes-sig-api-machinery/R4Iiz3rOYd0/O6aX15_AAQAJ

A universal, defaulting, strict decoder is needed for Component Config efforts. 
Future work aims to make strict behavior the default -- right now it's not implemented in a way that is performant enough for the apiserver.

This patch adds a deprecation notice for `NewCodecFactory()`.
For now, uses of this public constructor remain unchanged, and the tests are still exercising it.
It's been modified, however, to simply wrap the `WithOptions` version.

Tracking issue: https://github.com/kubernetes/kubernetes/issues/75426

**Edit:**
I've added another commit implementing option mutator funcs for the `NewCodecFactory`.
This allows us to keep using that constructor without deprecating it.
We can squash this into the previous commit if we agree we want this.
Note that it adds maintenance overhead to `CodecFactoryOptions` since each new option will require at least 2 new functions for booleans, and potentially many many more for other types.
The user can always implement their own function when calling the constructor, but this is clunky.

**Special notes for your reviewer**:
I'm not sure I like how verbose this incantation is:
```go
codecs.NewCodecFactoryWithOptions(
    scheme,
    codecs.CodecFactoryOptions{Pretty: true, Strict: true},
)
```
It's much longer than before.
If you have feedback on naming, it would be beneficial to make this not look so intimidating.

Note as well that the tests should be modified to exercise the non-Pretty and Strict paths.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

stealthybox committed:
Implement CodecFactoryOptions allowing clients to opt-in to Pretty encoders and Strict Decoders

This patch introduces a new CodecFactory constructor that's capable of building different collections of Serializers.
Note that this diverges from the current design where an initialized CodecFactory is a kitchen sink holding many forms of useful Serilaizers in a non-customizeable way.

This approach is a suggested solution from this thread:
https://groups.google.com/d/msg/kubernetes-sig-api-machinery/R4Iiz3rOYd0/O6aX15_AAQAJ

A universal, defaulting, strict decoder is needed for Component Config efforts.
Future work aims to make strict behavior the default -- right now it's not implemented in a way that is performant enough for the apiserver.

This patch adds a deprecation notice for `NewCodecFactory()`.
For now, uses of this public constructor remain unchanged, and the tests are still exercising it.
It's been modified, however, to simply wrap the `WithOptions` version.

stealthybox commented:
/wg component-standard

stealthybox commented:
/retest

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -92,15 +108,34 @@ type CodecFactory struct {
+	// Strict: configures all serializers in strict mode
+	Strict bool
+	// Pretty: includes a pretty serializer along with the non-pretty one
+	Pretty bool
```
should we add similar comment/details as in SerializerOptions, that Pretty is only for JSON encoding and Strict is only for decoding?

stealthybox replied:
I'm not sure.
Maintaining these docs is tougher if the behavior of `SerializerOptions` changes.

I'm fine to add more detail here -- it's good for library users.
Let me know what you think.

stealthybox replied:
I could add a note that refers to the SerializerOptions godoc?

neolit123 replied:
i'm leaning towards adding the notes, but let's see what the others think.

logicalhan replied:
Personally, I'd prefer the details here. I had to trace through the code to see why `CodecFactoryOptions` ignored the value for field `Pretty` for the yamlSerializer in `newSerializersForScheme`.

neolit123 reviewed (commented):
thanks @stealthybox 
/priority important-longterm
/retest

stewart-yu commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -47,30 +47,44 @@ type serializerType struct {
+		Framer:           json.Framer,
+		StreamSerializer: jsonSerializer,
+	}
+	if options.Pretty {
```
must we construct `jsonPrettySerializer` , rather than decided by `options.Pretty`?

neolit123 replied:
i think @stewart-yu is correct.

stealthybox replied:
The YAML one just leaves it defaulted.
This is the first thing I think to do to make `Pretty` optional.

We need to look at how the `serializerType`'s are used `newCodecFactory` though. Some logic might be missing here.

stealthybox replied:
@stewart-yu are you saying `jsonSerializerType.PrettySerializer` is always required even when `Pretty: false` ?

stewart-yu replied:
yes

caesarxuchao commented:
/assign @logicalhan

logicalhan commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -47,30 +47,44 @@ type serializerType struct {
+
+	yamlSerializer := json.NewSerializerWithOptions(
+		mf, scheme, scheme,
+		json.SerializerOptions{Yaml: true, Pretty: false, Strict: options.Strict},
```
Is it better to omit the pretty field here since it no-opts (according to `SerializerOptions` documentation)?

stealthybox replied:
My thinking here was that Pretty would be set to `false` to imply that it doesn't matter for YAML.
This is the trouble with implementing a single Options struct in Serializer... adding more doesn't sound better though.

I'm fine with omitting this if you think it's more appropriate.

stealthybox replied:
Also, @luxas suggests that we use option mutator funcs for `NewSerializer`, but the signature change will require a code refactor, so I'm not sure if it's worth doing.

I'm adding a patch to this PR to try out the mutator funcs for `NewCodecFactory` since the current signature does not have any options.

luxas commented:
/assign @liggitt @lavalamp @sttts 
for approval. @stealthybox can rebase later when the idea itself is approved.

stealthybox committed:
Extend NewCodecFactory() signature with options mutators

When no mutators are passed, the legacy behavior is implemented.
This allows code to continue using NewCodecFactory() without modifications.

This also reverts the deprecation notice from the previous commit.

stealthybox commented:
Updated PR description:
> I've added another commit implementing option mutator funcs for the NewCodecFactory.
This allows us to keep using that constructor without deprecating it.
We can squash this into the previous commit if we agree we want this.
Note that it adds maintenance overhead to CodecFactoryOptions since each new option will require at least 2 new functions for booleans, and potentially many many more for other types.
The user can always implement their own function when calling the constructor, but this is clunky.

stealthybox commented:
TODO:
- [ ] Update docstrings to make user discovery of `CodecFactoryOptions` / `SerializerOptions` behavior more clear/discoverable ( suggested by @logicalhan )

stealthybox commented:
PS: I'm unsure of the behavior of JSON when `CodeFactoryOptions.Pretty` is false.
The intermediary defaults a nil interface for `serializerType.PrettySerializer` -- this gets passed into a `runtime.SerializerInfo`. @stewart-yu 

This really needs a test suite.

luxas commented:
This looks ok to me!
Yes, tests would be good ;)

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -104,14 +122,70 @@ type CodecFactory struct {
+// TODO: allow other codecs to be compiled in?
+// TODO: accept a scheme interface
+func NewCodecFactory(scheme *runtime.Scheme, mutators ...CodecFactoryOptionsMutator) CodecFactory {
+	options := CodecFactoryOptions{}
```
@sttts suggested that the better way of doing this is by setting `Pretty=true` by default

stealthybox replied:
:heavy_check_mark:

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -104,14 +122,70 @@ type CodecFactory struct {
 // TODO: accept a scheme interface
-func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
-	serializers := newSerializersForScheme(scheme, json.DefaultMetaFactory)
+func NewCodecFactoryWithOptions(scheme *runtime.Scheme, options CodecFactoryOptions) CodecFactory {
```
make this private, at least for now

stealthybox replied:
removed this intermediate func :heavy_check_mark:

luxas commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -48,41 +48,57 @@ type serializerType struct {
-	serializer := protobuf.NewSerializer(scheme, scheme)
-	raw := protobuf.NewRawSerializer(scheme, scheme)
+func newSerializersForScheme(scheme *runtime.Scheme, mf json.MetaFactory, options CodecFactoryOptions) []serializerType {
+	jsonSerializer := json.NewSerializerWithOptions(
```
Instead of doing it this way, please keep the list, and just use the same variables like before, that you create conditionally

stealthybox replied:
reduced this :heavy_check_mark:

luxas reviewed (commented):
Goes in the right direction, but we'd like to see some small changes before merge
Pair-reviewed with @sttts

stealthybox committed:
Make Pretty always default and remove public WithOptions constructor

stealthybox committed:
Reduce []serializerType{} patch

stealthybox committed:
Test Strict CodecFactory option when enabled/disabled

stealthybox commented:
Tests are now pushed covering the Strict option. :+1:

> I'm unsure of the behavior of JSON when CodeFactoryOptions.Pretty is false.
The intermediary defaults a nil interface for serializerType.PrettySerializer -- this gets passed into a runtime.SerializerInfo. 

^ I looked into PrettySerializer -- it's correct to just leave it as `nil` if you don't want human-readable output.

This is currently ignored, though. (We can see this behavior in the package's existing tests, where the JSON is not prettified.)

The only place PrettySerializer is currently used in k/k is the apiserver where it actually reconstructs overwriting the non-pretty encoder:
[47e52d2: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go#L53-L55](https://github.com/kubernetes/kubernetes/blob/47e52d2/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go#L53-L55)

There's no current helper to easily access this pretty encoder when the option is enabled on the CodecFactory outside of the apiserver's request media-type negotiation.
We can add that in a future patch when it's needed.

stealthybox commented:
/milestone 1.16

@sttts @logicalhan WDYT about merging this soon?

sttts commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -104,14 +118,56 @@ type CodecFactory struct {
 
+// CodecFactoryOptions holds the options for configuring CodecFactory behavior
+type CodecFactoryOptions struct {
+	// Strict: configures all serializers in strict mode
```
no colon for godoc style

stealthybox replied:
:heavy_check_mark:

sttts commented:
A comment nit.

lgtm otherwise.

@liggitt this needs api approval (why?)

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -96,6 +108,8 @@ func newSerializersForScheme(scheme *runtime.Scheme, mf json.MetaFactory) []seri
 // CodecFactory provides methods for retrieving codecs and serializers for specific
 // versions and content types.
 type CodecFactory struct {
+	options CodecFactoryOptions
```
I'm not seeing where this is used... am I missing it?

stealthybox replied:
Good catch! :pray: 
This was used in an older implementation.
Removed now :heavy_check_mark:

liggitt commented:
> @liggitt this needs api approval (why?)

it doesn't... the comment in the description added api-change. any approver in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/OWNERS can approve

stealthybox committed:
Fix CodecFactoryOptions nits

stealthybox commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-gce-100-performance

stealthybox commented:
/test pull-kubernetes-e2e-gce

stealthybox commented:
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-gce-100-performance

stealthybox commented:
/test pull-kubernetes-e2e-gce-100-performance

stealthybox commented:
/test pull-kubernetes-e2e-gce-100-performance

stealthybox commented:
@liggitt @sttts -- addressed those comments :+1:

liggitt commented:
/lgtm
/approve
